### PR TITLE
Schedule compaction

### PR DESCRIPTION
**Pull request by paili0628**
implements #1717. 
I have not yet put the `schedule-compaction` pass into the default pipeline but I think it should be immediately after `static-promotion`, since I think it is only reasonable to compact static seqs that were promoted by the `static-promotion` pass, but I want to know what you think @calebmkim 

Update: Already inserted `schedule-compaction` into default pipeline after `static-promotion`.

On Monday we discussed that while the current version of `schedule-compaction` is adequate, it is worthwhile to extend this optimization to nested static seqs with `repeat`, `par`, `seq` and `while` blocks as children. To address the problem that we can only utilize the `go` signal of groups, @calebmkim suggested that instead of making a big group scheduling `go` signals, we use a `static par` cleverly to achieve the same effect.

### CODE REVIEW

File: calyx-opt/src/passes/schedule_compaction.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+use ir::{build_assignments, structure, Attributes, StaticTiming};
+use petgraph::{algo, graph::NodeIndex};
+use std::collections::HashMap;
+
+#[derive(Default)]
+/// for static seqs that are statically promoted by the compiler,
+/// aggressively compacts the execution schedule so that the execution
+/// order of control operators still respects data dependency
+/// Example: see tests/passes/schedule-compaction/schedule-compaction.rs
+
```
Comment (rachitnigam) [inline calyx-opt/src/passes/schedule_compaction.rs:16]: No space here

File: calyx-opt/src/passes/schedule_compaction.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                        let allow_start = schedule[node] + latency_map[node];
+                        if allow_start > start {
+                            start = allow_start;
+                        }
+                    }
+                    schedule.insert(i, start);
+
+                    let control = total_order[i].take().unwrap();
+
+                    match control {
```
Comment (rachitnigam) [inline calyx-opt/src/passes/schedule_compaction.rs:84]: Use `if`-`let` instead of matching on just one variant

**Comment by rachitnigam** [comment]
@paili0628 any updates on the state of this and what would it take to merge this? I think it would be good to summarize the discussion we had in the meeting here to describe the 3 steps it would take to get the complete schedule compaction we wanted.

**Comment by paili0628** [comment]
I'm currently debugging this pr. It seems that for some test cases the while loops are violating the bound constraints even though `schedule-compaction` does not seem to have changed the code 'visibly'.

**Comment by calebmkim** [comment]
Is it possible that it's not keeping track of dependencies across iterations of the while loop? For example:
```
while port {
    A; // reads from register. This read from A depends on the write to the register in C. 
    B;
    C; // writes to register. 
} 
```
I haven't looked at the code carefully; maybe you have covered this.

**Comment by paili0628** [comment]
The problem is that `while` loops with `comb group`s are replaced before `schedule-compaction`, leading to `schedule-compaction` messing up the scheduling of `cond` in some cases.

**Comment by calebmkim** [comment]
I see. I think we should probably try to handle this case? (Although I'm not sure how difficult that would be). 

I had a similar problem when I was trying to use graph coloring to reuse FSMs for static islands that we know wouldn't run at the same time. (In other words, when instantiating FSMs for static islands, I made it so that different static islands can use the same FSM as long as they're guaranteed not be active at the same time). The code is [here](https://github.com/cucapra/calyx/blob/590c1945de6ce993aae0928770e692bd7b1d7a21/calyx-opt/src/passes/compile_static.rs#L599). Looking back, the code is not necessarily the best, so I could envision factoring out some of this code into a separate analysis that this pass will use as well.

**Comment by paili0628** [comment]
I see what you're saying. I think the code in `compile-static` is definitely helpful. The only problem is that, I think the current `ReadWriteSet::port_write_set` function only takes in an `Iterator` of assignments. In order to obtain all the cells in its 'referred' groups, it will have to take in the component as well, which might affect how other passes or analyses use this function. This is just a potential difficulty and does not mean it's not doable.

**Comment by rachitnigam** [comment]
@paili0628 are there technical discussions we need to have to move this PR along? Let's try to wrap this up soon so we can start working on results

**Comment by paili0628** [comment]
There are no technical discussions. I only have to deal with fmt. Sorry about this.

**Comment by rachitnigam** [comment]
No worries!

**Comment by paili0628** [comment]
This is ready for performance evaluation. @calebmkim

**Comment by calebmkim** [comment]
Awesome thanks!

**Pull request merged**